source/common/http/http2/codec_impl.cc

```diff
@@ -456,6 +500,8 @@ int ConnectionImpl::onStreamClose(int32_t stream_id, uint32_t error_code) {
     }
 
     connection_.dispatcher().deferredDelete(stream->removeFromList(active_streams_));
+    nghttp2_session_consume(session_, stream_id, stream->unconsumed_bytes_);
```

Comment:
I would have thought so too.  I had a random moment of "oh but how awful would it be to debug things it doesn't?" and added a regression test   just in case.  (Http2CodecImplFlowControlTest.EarlyResetRestoresWindow) which fails without this consume.

You don't need to update stream windows on stream close, but the stream window influences the connection window and that isn't updated.  My first reaction was that was terrible.  My second reaction was that if a stream is "closed" as far as nghttp2 is concerned (all data is sent, fin is received) but the consumer has the data buffered, it really ought not update the connection level window until the consumer says to.  So yeah, necessary by design :-/